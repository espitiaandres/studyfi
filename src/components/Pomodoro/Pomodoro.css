.controlButtons {
    background-color: #1ED760;
    border: 2px solid #1ED760;
    padding: 3px 12px;
    margin: 0px 2px;
    outline: none;
}

.controlButtons:hover {
    opacity: 0.7;
}

.countdown {
    color: black;
    background-color: #1ED760;
    padding: 0.5px 2px 3.5px 2px;
    margin-left: 2px;
}

.pomodoroPanel {
    align-items: center;
    display: flex;
    justify-content: center;
}

.pomodoroHeader {
    align-items: center;
    color: white;
    display: flex;
    flex-direction: column;
    justify-content: center;
}

.controlButtons-pizzaslice {
    background-color: #1ED760;
    border: 2px solid #1ED760;
    padding: 3px 12px;
    margin: 0px 2px;
    outline: none;
}
