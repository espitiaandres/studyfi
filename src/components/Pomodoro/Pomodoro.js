//
//  Pomodoro.js
//  react-spotify-player
//
//  Created by Andres Espitia.
//  Copyright Â© 2020 Andres Espitia. All rights reserved.
//

import React, { useState, useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';
import { fas } from '@fortawesome/free-solid-svg-icons';
import { fab } from '@fortawesome/free-brands-svg-icons';
import { Link } from 'react-scroll';
import { holidaysColors } from '../../utils/holidays';
import moment from 'moment';
import 'animate.css';
import './Pomodoro.css';

const secondsInAMinute = 60;
const setTimeMinutes = 25;                                              // minutes
const setTimeSeconds = setTimeMinutes * secondsInAMinute - 1;           // seconds
const cooldownSetTimeMinutes = 5;
const cooldownSetTimeSeconds = cooldownSetTimeMinutes * secondsInAMinute;
let remainingTime = setTimeSeconds;
let minutesRemaining = setTimeMinutes < 10 ? "0" + setTimeMinutes.toString() : setTimeMinutes.toString();
let secondsRemaining = "00";
let timerRunning = false;
let pomodoroText = "press on the play button below to begin";
let interval;
let cooldownInterval;

library.add(fas, fab);

const Pomodoro = ({ user, userProfile, season }) => {
    let seasonStyling = season ? "seasonStyling" : "";
    let seasonStylingAlt = season ? "seasonStylingAlt" : "";

    const [playButton, setPlayButton] = useState(false);
    const [restartButton, setRestartButton] = useState(false);
    const [pizzaButton, setPizzaButton] = useState(false);
    const [greeting, setGreeting] = useState('');

    useEffect(() => {
        const now = moment();
        getGreetingSuffix(now);        
    }, [])

    const getGreetingSuffix = (now) => {
        const currentHour = now.format("HH");
        const afternoonTime = 12;
        const eveningTime = 18;

        if (currentHour > eveningTime) {
            setGreeting('Good evening');
        } else if (currentHour > afternoonTime) {
            setGreeting('Good afternoon');
        } else {
            setGreeting('Good morning')
        }
    }

    const countdown = () => {
        setRestartButton(false);
        if (playButton) {
            return;
        }
        remainingTime = setTimeSeconds;
        pomodoroText = season ? "spookspookspookspookspook" : "WorkWorkWorkWorkWork";
        remainingTime++;
        updatedFormattedTimeString(remainingTime);
        timerRunning = true;
        interval = setInterval(() => {
            if (remainingTime > 0) {
                remainingTime--;
                updatedFormattedTimeString(remainingTime);
                timerRunning = true;
            } else {
                clearInterval(interval);
                countdownCooldown();
            }
        }, 1000);
        setPlayButton(true);
        setPizzaButton(true);
    }

    const countdownCooldown = () => {
        remainingTime = cooldownSetTimeSeconds;
        pomodoroText = season ? "2sp00ky4me" : "RestRestRestRestRest";
        remainingTime++;
        cooldownInterval = setInterval(() => {
            if (remainingTime > 0) {
                remainingTime--;
                updatedFormattedTimeString(remainingTime);
            } else {
                clearInterval(cooldownInterval);
                countdown();
            }
        }, 1000)
    }

    const clearCountdown = () => {
        setPlayButton(false);
        setPizzaButton(false);
        if (restartButton) {
            return;
        }
        pomodoroText = season ? "2sp00ky4me" : "RestRestRestRestRest";
        clearInterval(interval);
        remainingTime = setTimeSeconds + 1;
        timerRunning = false;
        updatedFormattedTimeString(remainingTime);
        setRestartButton(true);
    }

    const updatedFormattedTimeString = (remainingTime) => {
        minutesRemaining = Math.floor(remainingTime / secondsInAMinute);
        let baseMinutes = minutesRemaining * secondsInAMinute;
        if (minutesRemaining < 10) {
            minutesRemaining = minutesRemaining === 0 ? "00" : "0" + minutesRemaining.toString();
        }
        secondsRemaining = remainingTime - baseMinutes;
        if (secondsRemaining < 10) {
            secondsRemaining = secondsRemaining === 0 ? "00" : "0" + secondsRemaining.toString();
        }
    }

    const colorSchema = holidaysColors.halloween;
    const seasonIconOuter = colorSchema.iconOuter; 
    const seasonIconInner = colorSchema.iconInner;
    const seasonIconTimer = colorSchema.iconTimer;
    const seasonColor = colorSchema.color;
    const seasonColorAlt = colorSchema.colorAlt;

    return (
        <div className="pomodoroHeader">
            <h1 className="greetingText">
                {greeting}, 
                <a className={`userProfile userProfile${seasonStyling}`} href={userProfile} target="_blank" rel="noopener noreferrer">
                    {` ${user}`}
                </a>                
            </h1>

            <div className="pomodoroHeaderTitle">
                <div className={`iconsSpacing ${seasonStyling}Icons`}>
                    <FontAwesomeIcon icon={season ? seasonIconOuter : ["fas", "pizza-slice"]} />
                </div>
                <div className={`iconsSpacing ${seasonStylingAlt}Icons`}>
                    <FontAwesomeIcon icon={season ? seasonIconInner : ["fas", "utensils"]} />
                </div>
                <h1 className="pomodoroHeaderTitleText">{season ? "pomodorooooo" : "pomodoro"}</h1>
                <div className={`iconsSpacing ${seasonStylingAlt}Icons`}>
                    <FontAwesomeIcon icon={season ? seasonIconInner : ["fas", "utensils"]} />
                </div>
                <div className={`iconsSpacing ${seasonStyling}Icons`}>
                    <FontAwesomeIcon icon={season ? seasonIconOuter : ["fas", "pizza-slice"]} />
                </div>
            </div>

            <div className="pomodoroHeaderTitle">
                <div className={`iconsSpacing ${seasonStyling}Icons`}>
                    <FontAwesomeIcon icon={season ? seasonIconTimer : ["fas", "clock"]} />
                </div>
                <div className={`countdown ${seasonStyling}AltIcons`}>
                    {minutesRemaining}:{secondsRemaining}
                </div>
                <div className={`iconsSpacing ${seasonStyling}Icons`}>
                    <FontAwesomeIcon icon={season ? seasonIconTimer : ["fas", "clock"]} />
                </div>
            </div>
            <p>{pomodoroText}</p>
            <div className="pomodoroPanel">
                <button className="controlButtons" onClick={countdown} disabled={pizzaButton}>
                    {season 
                    ? 
                    <FontAwesomeIcon icon={seasonIconOuter} style={{ color: seasonColor }} /> 
                    : 
                    <FontAwesomeIcon icon={["fas", "pizza-slice"]} style={{ color: "#FFF" }} />
                    }
                </button>
                <button className="controlButtons" onClick={clearCountdown} disabled={restartButton}>
                    <FontAwesomeIcon icon={["fas", "undo"]}  style={{ color: season ? seasonColorAlt : "#FFF" }}/>
                </button>
                <button className="controlButtons" onClick={countdown} disabled={playButton}>
                    { 
                    !timerRunning 
                    ? 
                    <FontAwesomeIcon icon={["fas", "play"]} style={{ color: season ? seasonColor : "#FFF" }} /> 
                    : 
                    <FontAwesomeIcon icon={["fas", "pause"]} style={{ color: season ? seasonColor : "#FFF" }} /> 
                    }
                </button>
            </div>

            <div className="icon-scroll animate__animated animate__bounce animate__infinite animate__slow">
                <Link to="topsongs" spy={true} smooth={true} duration={400}>
                    <FontAwesomeIcon icon={"chevron-down"} size="2x" />
                </Link>
            </div>
        </div>
    )
}

export default Pomodoro
