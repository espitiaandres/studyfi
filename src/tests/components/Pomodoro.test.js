//
//  Pomodoro.test.js
//  react-spotify-player
//
//  Created by Andres Espitia.
//  Copyright Â© 2020 Andres Espitia. All rights reserved.
//

import React from 'react';
import { shallow } from 'enzyme';
import Pomodoro from '../../components/Pomodoro/Pomodoro';

const pomodoroPropsTest = {
    user: "testUser",
    userProfile: "https://open.spotify.com/user/testUser",
    season: true
}

test('should render Pomodoro component correctly', () => {
    const wrapper = shallow(<Pomodoro />);
    expect(wrapper).toMatchSnapshot();
});

test('should render Pomodoro component correctly with data', () => {
    const wrapper = shallow(
        <Pomodoro 
            user={pomodoroPropsTest.user}
            userProfile={pomodoroPropsTest.userProfile}
            season={pomodoroPropsTest.season}
        />
    );
    expect(wrapper).toMatchSnapshot();
});
