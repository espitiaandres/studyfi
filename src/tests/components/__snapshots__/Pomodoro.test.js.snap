// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Pomodoro component correctly 1`] = `ShallowWrapper {}`;

exports[`should render Pomodoro component correctly with data 1`] = `ShallowWrapper {}`;
